### ISSUE TITLE
Smooth grid edges

### ISSUE THREAD

**Issue by rpachaly** [FLO-2DSoftware/FLO-2DMapCrafter#23]
Find a way to smooth the grid edges like Mapper

![image](https://github.com/FLO-2DSoftware/FLO-2DMapCrafter/assets/39889306/ac02dea5-f7ee-430a-8f4d-d7b7f6a4d8a4)

![image](https://github.com/FLO-2DSoftware/FLO-2DMapCrafter/assets/39889306/ed01a34a-d256-4651-a5c9-483ccd780e24)

**Comment by rpachaly** [comment]
@FLO-2DKaren 
Hi Karen. Do you have any idea on how to do this on QGIS? I'm trying to find an efficient way to do the same thing that mapper does but nothing seems to smooth the edges the way mapper does. 

I need to run a couple of algorithms to get the same mapper result. It will take around 3-5 seconds per map I think.

**Comment by FLO-2DKaren** [comment]
Did you use raster contour tool?

**Comment by rpachaly** [comment]
![image](https://github.com/FLO-2DSoftware/FLO-2DMapCrafter/assets/39889306/ecbbef07-2ce5-4414-9ad0-5cd0f22d3c64)

To get this kind of result, I follow these steps:
1. Create the "Contour Polygons" from the raster
2. Create the "Raster pixel to points"
3. Create the "Concave hull (k-nearest neighbor)" using the "Raster pixel to points" -> to remove the square edges
4. Apply the "Smooth" for better visualization using the "Concave hull (k-nearest neighbor)" -> it is not necessary
5. Clip the "Contour Polygons" with "Smooth" to remove the edges. 

It is not efficient but it works.

**Comment by rpachaly** [comment]
It works fine for continuous data, but for categorized data (e.g. hazard maps) it is bad.

**Comment by FLO-2DKaren** [comment]
Maybe we should see what the Mapper code is doing. I’ll ask JJ to give us some help.

**Comment by FLO-2DKaren** [comment]
@FLO-2DJJ   hi JJ. Can you tell us how Mapper builds contour maps?  If you follow this conversion you’ll get an idea of what’s happening with the raster maps.

**Comment by rpachaly** [comment]
> hould see what the Mapper code is doing. I’ll ask JJ to give us some help.

Great idea! If I have to code everything, I'll have something to be based on.

**Comment by FLO-2DJJ** [comment]
Yes. I've been following this. Great work Robson. Tomorrow I'll see if
there is something that can be done according to my experience with Mapper.
It uses a different approach but could show some light on the issue.

**Comment by FLO-2DJJ** [comment]
>     1. Create the "Contour Polygons" from the raster
>     2. Create the "Raster pixel to points"
>     3. Create the "Concave hull (k-nearest neighbor)" using the "Raster pixel to points" -> to remove the square edges
>     4. Apply the "Smooth" for better visualization using the "Concave hull (k-nearest neighbor)" -> it is not necessary
>     5. Clip the "Contour Polygons" with "Smooth" to remove the edges.

@rpachaly hi Robson, can I see the code you are using here. Do you have a branch with this code so that I can see it? I t doesn't seem to be in the "new_mapping" branch.

**Comment by rpachaly** [comment]
@FLO-2DJJ 
Hi JJ, 

No, this is not coded yet. This is just the QGIS native algorithms that I intent to use. If you think this is a good approach, I'll code this way. Otherwise, feel free to suggest other approaches.

**Comment by FLO-2DJJ** [comment]
Ok. No problem. Currently I'm looking at the old Mapper code to see the
approach. The code itself is obsolete and no longer supported but it could
show a direction to follow.
Thanks.

**Comment by FLO-2DJJ** [comment]
To create filled contours (shaded contours) in the old Mapper, like this Maximum Water Surface Elevation
![image](https://github.com/FLO-2DSoftware/FLO-2DMapCrafter/assets/20424575/5e7f41d7-cc66-48ca-bab7-2deb931b2735)

 it doesn't create contour lines first. Instead it creates a 2D rectangular matrix (280 x 222 in this example) with the value of the variable in each cell (or a very small negative value to indicate novalue). Then it calls a SimplePlot function  (SimplePlot was the name software used) to create the the plot. 
Maybe we could try a similar approach in QGIS. I have not found it yet.

**Comment by rpachaly** [comment]
Thanks JJ! This is a good direction to follow. I don't think that creating the contours is a good approach, specially for categorized data.

**Comment by rpachaly** [comment]
The GRASS r.resamp.inter algorithm interpolate the raster values and create a very good internal result. But the edges are still grid-like and it misses some areas close to the raster's edges. 

![image](https://github.com/FLO-2DSoftware/FLO-2DMapCrafter/assets/39889306/b1262edd-9772-40eb-87ae-4b0246887f53)

**Comment by rpachaly** [comment]
This plugin is also good. It creates contour polygons from points. https://github.com/ccrook/QGIS-Contour-Plugin

**Comment by FLO-2DJJ** [comment]
Interesting. I'm going to try creating cell centroids and continue
from there.

**Comment by FLO-2DJJ** [comment]
Yes Robson, the https://github.com/ccrook/QGIS-Contour-Plugin plugin
produces acceptable results. Let's try its several options to see if we can
adopt it.

This is what I got when using the GROUND_ELEVATION raster layer:
[image: imagen.png]


   - From the GROUND_ELEVATION create a Vector points layer using *Processing
   Toolbox. Vector Creation. Raster pixels to points*:


[image: imagen.png]

It creates a 'Vector points' layer.

   - Symbolize it:

[image: imagen.png]

[image: imagen.png]

   - Use the  aforementioned plugin with its defaults when you select the
   'Vectors points' layer:

[image: imagen.png]

[image: imagen.png]


I suppose something has to be done to the initial raster or points layer to
eliminate the cells (or points) with no value to obtain something similar
to the old Mapper:

[image: unnamed-1.png]


Juan Jose Rodriguez
FLO-2D Software
www.flo-2d.com
***@***.***

**Comment by FLO-2DJJ** [comment]
Yes Robson, the https://github.com/ccrook/QGIS-Contour-Plugin plugin produces acceptable results. Let's try its several options to see if we can adopt it.

- This is what I got when using the GROUND_ELEVATION raster layer:

![imagen](https://github.com/FLO-2DSoftware/FLO-2DMapCrafter/assets/20424575/7af4c7a4-580b-4dba-8470-9a880482aa16)

- From the GROUND_ELEVATION create a Vector points layer using **Processing Toolbox. Vector Creation. Raster pixels to points**:

![imagen](https://github.com/FLO-2DSoftware/FLO-2DMapCrafter/assets/20424575/39a4d115-0fa2-44e3-a388-83796292199d)

It creates a 'Vector points' layer.

- Symbolize it:

![imagen](https://github.com/FLO-2DSoftware/FLO-2DMapCrafter/assets/20424575/52d31d69-ca95-48a6-bbfd-2513b4ac33f5)
![imagen](https://github.com/FLO-2DSoftware/FLO-2DMapCrafter/assets/20424575/7b53fe0a-1544-44a0-86a4-4c688483a5c5)

- Use the  aforementioned plugin with its defaults when you select the 'Vectors points' layer:

![imagen](https://github.com/FLO-2DSoftware/FLO-2DMapCrafter/assets/20424575/fb5697ed-f418-467b-ba0a-43179cfdd34e)
![imagen](https://github.com/FLO-2DSoftware/FLO-2DMapCrafter/assets/20424575/da6b023d-7ce6-4fa5-9bed-99e8f9321e98)

- I suppose something has to be done to the initial raster or points layer to eliminate the cells (or points) with no value to obtain something similar to the old Mapper:

![279021435-5e7f41d7-cc66-48ca-bab7-2deb931b2735](https://github.com/FLO-2DSoftware/FLO-2DMapCrafter/assets/20424575/2ff1f157-d69f-4e52-87b7-f43c1bc77cad)

**Comment by rpachaly** [comment]
That looks good! I'm think that the best option so far is to understand the Contour Plugin logic and adapt it to MapCrafter. I was worried about the processing time to create a map, but Mapper also takes some seconds to generate a Shaded Contour map.

**Comment by FLO-2DKaren** [comment]
I suggest we let the end users figure this out for now.  Not many people request contours and those who do should be able to figure it out with the QGIS toolbox.  Let's not spend more time on this.  We have more important tasks.
